sos/report/plugins/mellanox_firmware.py

```diff
@@ -0,0 +1,140 @@
+        We will only enable the plugin if there is a
+        Mellanox Technologies network adapter
+        """
+        lspci = self.exec_cmd("lspci -D -d 15b3::0200")
+        if lspci['status'] != 0:
+            return False
+        return self.MLNX_STRING in lspci['output']
+
+    def collect(self):
+        if not self.get_option('allow_system_changes'):
+            return
+
+        """
+        Run only if mft package is installed.
+        flint is available from the mft package.
+        """
+        co = self.exec_cmd('flint --version')
+        if co['status'] != 0:
+            return
+
+        co = self.collect_cmd_output('mst start')
+        if co['status'] != 0:
+            return
+
+        self.collect_cmd_output('mst cable add')
+        self.collect_cmd_output("mst status -v", timeout=10)
+        self.collect_cmd_output("mlxcables", timeout=10)
+        co = os.listdir("/dev/mst")
+        mlxcables = []
+        for device in co:
+            if 'cable' in device:
+                mlxcables.append(device)
+        for mlxcable in mlxcables:
+            self.collect_cmd_output(f"mlxcables -d {mlxcable} --DDM",
+                                    timeout=10)
+            self.collect_cmd_output(f"mlxcables -d {mlxcable} --dump",
+                                    timeout=10)
+        self.collect_cmd_output("mst stop", changes=True)
+
+    def setup(self):
+        # Get all devices which have the vendor Mellanox Technologies
+        devices = []
+        device_list = self.collect_cmd_output('lspci -D -d 15b3::0200')
+        """
+        Will return a string of the following format:
+        0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+        Family
+        """
+        if device_list['status'] != 0:
+            # bail out if there no Mellanox PCI devices
+            return
+
+        for line in device_list["output"].splitlines():
+            """
+            Should return 0000:08:00.0
+            from the following string
+            0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+            Family
+            """
+            devices.append(line[0:8]+'00.0')
+
+        devices = set(devices)
+
+        """
+        # Mft package is present if OFED is installed
+        # mstflint package is part of the distro and can be installed.
+        """
+        commands = []
+
+        # mft package is installed if flint command is available
+        co = self.exec_cmd('flint --version')
+        if co['status'] != 0:
+            """
+            mstflint package commands
+            the commands do not support position independent arguments
+            """
+            commands = [
+                ["mstconfig -d ", " -e q"],
+                ["mstflint -d ", " dc"],
+                ["mstflint -d ", " q"],
+                ["mstreg -d ", " --reg_name ROCE_ACCL --get"],
+                ["mstlink -d ", ""],
+            ]
+        else:
```

Comment:
This is confusing / contradicting with "if `flint --version` returns nonzero status, return from `collect` method".

Reply:
In collect we bail out if we had a nonzero status because mft is not installed.

This else branch is on zero status and we run the some commands since mft is installed.

Maybe it would be less confusing if check [here](https://github.com/sosreport/sos/pull/3407/files#diff-b3d320e633b5fb0c0545076c63fe766dd33650c2b7b6c911533b55194a8de55dR100) for zero status?

Reply:
Ah I understand, depending on the (non)zero status, you want to (not) collect different stuff (and in different methods, very strictly saying).